### PR TITLE
OpenSslEngine option to wrap/unwrap multiple packets per call

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The JDK SSLEngine documentation says that a call to wrap/unwrap "will attempt to consume one complete SSL/TLS network packet" [1]. This limitation can result in thrashing in the pipeline to decode and encode data that may be spread amongst multiple SSL/TLS network packets.
ReferenceCountedOpenSslEngine also does not correct account for the overhead introduced by each individual SSL_write call if there are multiple ByteBuffers passed to the wrap() method.

Modifications:
- OpenSslEngine and SslHandler supports a mode to not comply with the limitation to only deal with a single SSL/TLS network packet per call
- ReferenceCountedOpenSslEngine correctly accounts for the overhead of each call to SSL_write
- SslHandler shouldn't cache maxPacketBufferSize as aggressively because this value may change before/after the handshake.

Result:
OpenSslEngine and SslHanadler can handle multiple SSL/TLS network packet per call.

[1] https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/SSLEngine.html

### CODE REVIEW

**Comment by Scottmitch** [comment]
- This depends upon https://github.com/netty/netty-tcnative/pull/233
- Once https://github.com/netty/netty/pull/6360 is pulled in I will update to account for multiple packets in the buffer.
- Discovered https://github.com/netty/netty/issues/6348 while developing this. TODOs will be addressed as part of that issue.

File: codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java [hunk 1]
```diff
@@ -206,7 +206,7 @@ protected ByteBuf internalBuffer() {
     }
 
     @Override
-    public final void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java]: @Scottmitch I would love to not do this as this will basically make the `handlerRemoved0` redundant.

File: handler/src/main/java/io/netty/handler/ssl/OpenSslClientContext.java [hunk 2]
```diff
@@ -181,10 +179,9 @@ public OpenSslClientContext(File trustCertCollectionFile, TrustManagerFactory tr
...
-                                KeyManagerFactory keyManagerFactory, Iterable<String> ciphers,
-                                CipherSuiteFilter cipherFilter, ApplicationProtocolConfig apn, String[] protocols,
-                                long sessionCacheSize, long sessionTimeout, boolean enableOcsp)
-            throws SSLException {
+                         KeyManagerFactory keyManagerFactory, Iterable<String> ciphers,
+                         CipherSuiteFilter cipherFilter, ApplicationProtocolConfig apn, String[] protocols,
+                         long sessionCacheSize, long sessionTimeout, boolean enableOcsp) throws SSLException {
         super(ciphers, cipherFilter, apn, sessionCacheSize, sessionTimeout, SSL.SSL_MODE_CLIENT, keyCertChain,
                 ClientAuth.NONE, protocols, false, enableOcsp);
         boolean success = false;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslClientContext.java]: as this seems to be only formatting can you revert this ?

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java [hunk 3]
```diff
@@ -275,7 +275,7 @@ public String run() {
         try {
             synchronized (ReferenceCountedOpenSslContext.class) {
                 try {
-                    ctx = SSLContext.make(SSL.SSL_PROTOCOL_ALL, mode);
+                    ctx = SSLContext.make(SSL.SSL_PROTOCOL_TLS, mode);
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java]: @Scottmitch this seems to be not related, please revert

**Comment by normanmaurer** [comment]
@Scottmitch in general I would prefer if we could have two different PRs. One which fixes the "ReferenceCountedOpenSslEngine correctly accounts for the overhead of each call to SSL_write" and the other which introduce the new "feature". Can you please split it up ?

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 4]
```diff
@@ -988,6 +1013,30 @@ private boolean ignoreException(Throwable t) {
         return false;
     }
 
+    @Override
+    public final void handlerRemoved(final ChannelHandlerContext ctx) throws Exception {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: See above

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java [hunk 5]
```diff
@@ -566,7 +572,7 @@ public void operationComplete(Future<Channel> future) {
         }
     }
 
-    @Test(timeout = 30000)
+    @Test(timeout = 60000)
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: Why was this needed ?

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java [hunk 6]
```diff
@@ -601,33 +615,37 @@ private static void compositeBufSizeEstimationGuaranteesSynchronousWrite(
         Channel cc = null;
         try {
             final Promise<Void> donePromise = group.next().newPromise();
-            final int expectedBytes = 469 + 1024 + 1024;
+            // The goal is to provide the SSLEngine with many ByteBuf components to ensure that the overhead for wrap
+            // is correctly accounted for on each component.
+            final int numComponents = 150;
+            final int desiredBytes = 16384;
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: can you add a comment why this value is chosen (like SSL record size)

File: codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java (hunk 1 above)
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java]: The issue is the buffer may be released by this method prematurely due to re-entry type issues. We have a few options on how to approach this but child classes not having the opportunity to control when the buffer released is a problem. We could deprecate `handlerRemoved0`, or do you have a more preferred approach?

File: handler/src/main/java/io/netty/handler/ssl/OpenSslClientContext.java (hunk 2 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslClientContext.java]: done

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java (hunk 3 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java]: done

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java (hunk 5 above)
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: This test now does 4x more work than it does before and the time constraints have changed. The point of the timeout is to catch "hung" tests and not performance issues and so to avoid false positives the timeout was doubled.

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java (hunk 6 above)
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: done

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 4 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: duplicate comment ... lets discuss above

**Comment by normanmaurer** [comment]
@Scottmitch also please update the pr title and description

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java (hunk 5 above)
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: ok sounds good.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 7]
```diff
@@ -853,24 +921,38 @@ public final SSLEngineResult unwrap(
...
-                    int pendingEncryptedBytes = min(packetLength, remaining);
-                    ByteBuf bioWriteCopyBuf = writeEncryptedData(src, pendingEncryptedBytes);
                     try {
-                        readLoop:
-                        for (; dstsOffset < dstsEndOffset; ++dstsOffset) {
+                        for (;;) {
                             ByteBuffer dst = dsts[dstsOffset];
                             if (!dst.hasRemaining()) {
                                 // No space left in the destination buffer, skip it.
+                                if (++dstsOffset >= dstsEndOffset) {
```
Comment (netkins) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: MAJOR Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS134)

File: pom.xml [hunk 8]
```diff
@@ -263,7 +263,7 @@
     <!-- Fedora-"like" systems. This is currently only used for the netty-tcnative dependency -->
     <os.detection.classifierWithLikes>fedora</os.detection.classifierWithLikes>
     <tcnative.artifactId>netty-tcnative</tcnative.artifactId>
-    <tcnative.version>2.0.1.Final</tcnative.version>
+    <tcnative.version>2.0.2.Final-SNAPSHOT</tcnative.version>
```
Comment (Scottmitch) [inline pom.xml]: reminder: this must be updated

**Comment by Scottmitch** [comment]
@normanmaurer - I have addressed all of your comments. PTAL

File: handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java [hunk 9]
```diff
@@ -57,9 +57,9 @@
 import java.nio.ByteBuffer;
 import java.nio.channels.ClosedChannelException;
 import java.security.KeyStore;
+import java.security.Provider;
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SSLEngineTest.java]: revert ?

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 10]
```diff
@@ -255,10 +249,20 @@ protected void deallocate() {
...
+
+            // setMode may impact the overhead.
+            maxWrapOverhead = SSL.getMaxWrapOverhead(ssl);
         } catch (Throwable cause) {
             SSL.freeSSL(ssl);
             PlatformDependent.throwException(cause);
         }
+        this.maxWrapOverhead = maxWrapOverhead;
+        // maxWrapBufferSize must be set after maxWrapOverhead because there maybe a dependency on this value.
+        maxWrapBufferSize = jdkCompatibilityMode ? maxEncryptedPacketLength() : maxEncryptedPacketLength() << 4;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: can you just add a comment why `<< 4` ?

**Comment by Scottmitch** [comment]
>  I would like to delay pulling this in until we did another netty release

sgtm

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 10 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: done ... a bit of a magic number ... the comment says we can make this configurable in the future if necessary. for now I think its fine to start with enough space for 16 TLS packets per wrap.

File: codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java [hunk 11]
```diff
@@ -466,7 +495,7 @@ protected void decodeLast(ChannelHandlerContext ctx, ByteBuf in, List<Object> ou
         if (in.isReadable()) {
             // Only call decode() if there is something left in the buffer to decode.
             // See https://github.com/netty/netty/issues/4386
-            decode(ctx, in, out);
+            decode0(ctx, in, out);
         }
     }
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java]: sure ... this PR will depend upon it but let me abstract it out.

Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java]: see https://github.com/netty/netty/pull/6728

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 12]
```diff
@@ -469,6 +469,49 @@ private int readPlaintextData(final ByteBuffer dst) {
...
+
+    /**
+     * Visible only for testing!
+     */
+    final int maxEncryptedPacketLength() {
+        return maxWrapOverhead + MAX_PLAINTEXT_LENGTH;
+    }
+
+    final int calculateMaxLengthForWrap(int plaintextLength, int numComponents) {
+        return (int) min(maxWrapBufferSize, plaintextLength + (long) maxWrapOverhead * numComponents);
```
Comment (netkins) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: CRITICAL Inconsistent synchronization of io.netty.handler.ssl.ReferenceCountedOpenSslEngine.maxWrapOverhead; locked 66% of time [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIS2_INCONSISTENT_SYNC)

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 13]
```diff
@@ -469,6 +469,65 @@ private int readPlaintextData(final ByteBuffer dst) {
...
+        return maxWrapOverhead + MAX_PLAINTEXT_LENGTH;
+    }
+
+    /**
+     * This method is intentionally not synchronized, only use if you know you are in the EventLoop
+     * thread and visibility on {@link #maxWrapBufferSize} and {@link #maxWrapOverhead} is achieved
+     * via other synchronized blocks.
+     */
+    final int calculateMaxLengthForWrap(int plaintextLength, int numComponents) {
+        return (int) min(maxWrapBufferSize, plaintextLength + (long) maxWrapOverhead * numComponents);
```
Comment (netkins) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: CRITICAL Inconsistent synchronization of io.netty.handler.ssl.ReferenceCountedOpenSslEngine.maxWrapOverhead; locked 83% of time [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIS2_INCONSISTENT_SYNC)

**Comment by Scottmitch** [comment]
@normanmaurer - any more comments? this currently depends upon `tcnative 2.0.2.Final-SNAPSHOT` ... do you want me to wait to merge until after a release?

**Comment by normanmaurer** [comment]
@Scottmitch no more comments... let me release tcnative first . I will do this over the next days

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 13 above)
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @Scottmitch just wonder if this is something we need to investigate

Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: let me know if you think there is a problem. The intention is not to necessarily to protect this variable in the synchronization block because it is mostly accessed inside the EventLoop (except maybe for `OpenSslSession#getPacketBufferSize()`).

**Comment by blucas** [comment]
@Scottmitch what is the latest on this PR?

**Comment by Scottmitch** [comment]
@blucas - We were just waiting for a release until we pulled this in. I will pull in shortly.

**Comment by Scottmitch** [comment]
4.1 (f7b3caedd) 4.0 (c68a6adce)
